src/stores/widgets/StopGapWidget.ts

```diff
@@ -391,5 +428,31 @@ export class StopGapWidget extends EventEmitter {
         if (!this.started) return;
         WidgetMessagingStore.instance.stopMessaging(this.mockWidget);
         ActiveWidgetStore.delRoomId(this.mockWidget.id);
+
+        if (MatrixClientPeg.get()) {
+            MatrixClientPeg.get().off('event', this.onEvent);
+            MatrixClientPeg.get().off('Event.decrypted', this.onEventDecrypted);
+        }
+    }
+
+    private onEvent = (ev: MatrixEvent) => {
+        if (ev.isBeingDecrypted() || ev.isDecryptionFailure()) return;
+        if (ev.getRoomId() !== ActiveRoomObserver.activeRoomId) return;
```

Comment:
I would feel a lot better if this checked against the room that the widget was in rather than relying on it being the active room by definition: I can easily imagine a race between the active room changing and the widgets for the old room unmounting causing events from one room to leak into the widget of another.

Reply:
I also have a few questions on why it only feeds events to the widget as they arrive if the widget is mounted and we're viewing the room, but I think those are probably more questions for the MSC, since that is indeed what it says.

Reply:
Like events added before the widget was opened but might still be needed by the widget?

Reply:
Like events added before the widget was opened but might still be needed by the widget?

Reply:
Yeah - or that you'd be viewing a room with a widget, then you briefly switch to another and then an event comes in that would have caused the widget to do a thing, so it never does the thing - it seems like it would limit the uses of the API quite a lot.

Reply:
hmm, fair. The MSC is probably the right place for that, though the rationale is it trying to balance the idea of privacy as much as possible with widget functionality. The question then becomes how far back the widget receives messages for.